### ISSUE TITLE
VB can't split long lines at .

### ISSUE THREAD

**Issue by paul1956** [dotnet/vblang#489]
In the following code I want to split long lines at "." like I can in c#. If I split the lines the same way they are split in C# I get Error    BC30157 Leading '.' or '!' can only appear inside a 'With' statement.   

```
    Private Shared Iterator Function GenerateNewScript(ifBlock As MultiLineIfBlockSyntax) As IEnumerable(Of SyntaxNode)
        Dim ifStatement As IfStatementSyntax = GenerateInvertedIfStatement(ifBlock.IfStatement)
        Yield SyntaxFactory.MultiLineIfBlock(ifStatement).WithStatements(GenerateNewTrueStatement(ifBlock.ElseBlock.Statements)).WithLeadingTrivia(ifBlock.GetLeadingTrivia()).WithTrailingTrivia(ifBlock.GetTrailingTrivia()).WithAdditionalAnnotations(Formatter.Annotation)

        For Each stmt As StatementSyntax In ifBlock.Statements
            Yield stmt.WithAdditionalAnnotations(Formatter.Annotation)
        Next
    End Function
```

Desired code formatted like in C# below.

```
        static IEnumerable<SyntaxNode> GenerateNewScript(MultiLineIfBlockSyntax ifBlock)
        {
            var ifStatement = GenerateInvertedIfStatement(ifBlock.IfStatement);
            yield return SyntaxFactory.MultiLineIfBlock(ifStatement)
                .WithStatements(GenerateNewTrueStatement(ifBlock.ElseBlock.Statements))
                .WithLeadingTrivia(ifBlock.GetLeadingTrivia())
                .WithTrailingTrivia(ifBlock.GetTrailingTrivia())
                .WithAdditionalAnnotations(Formatter.Annotation);

            foreach (var stmt in ifBlock.Statements)
            {
                yield return stmt.WithAdditionalAnnotations(Formatter.Annotation);
            }
        }
```

**Comment by gafter** [comment]
Transferred here from Roslyn by request at https://github.com/dotnet/vblang/issues/232#issuecomment-573068652

**Comment by Hamenopi** [comment]
In vb, put dot at end of line for method chaining

**Comment by AdamSpeight2008** [comment]
In most cases you can use `implicit line continuation` after a dot.

**Comment by pricerc** [comment]
> In most cases you can use `implicit line continuation` after a dot.

Yes, but if you're used to SQL (borrowing from @paul1956's original post)

```VB
dim result = SyntaxFactory.MultiLineIfBlock(ifStatement)
                .WithStatements(GenerateNewTrueStatement(ifBlock.ElseBlock.Statements))
                .WithLeadingTrivia(ifBlock.GetLeadingTrivia())
                .WithTrailingTrivia(ifBlock.GetTrailingTrivia())
                .WithAdditionalAnnotations(Formatter.Annotation)
```

would allow you to (e.g. when debugging)

```VB
dim result = SyntaxFactory.MultiLineIfBlock(ifStatement)
                .WithStatements(GenerateNewTrueStatement(ifBlock.ElseBlock.Statements))
'                .WithLeadingTrivia(ifBlock.GetLeadingTrivia())
'                .WithTrailingTrivia(ifBlock.GetTrailingTrivia())
                .WithAdditionalAnnotations(Formatter.Annotation)
```


The period at the start of the line tells one that it's continuing from the one before. And it's easier to see (easier for the human eye to parse) than a trailing period.

The biggest win I see for the proposal is from this scenario - improving productivity by allowing a developer to quick include and exclude code to speed up a debugging process by breaking things down.

The aesthetic of leading or trailing period is subjective, although for me, I'd prefer them (and commas, for that matter) to be leading the next line.

SQL's parser can manage this without* having statement delimiters, so we know it's possible. (*note: yes, I know that in SQL the semi-colon is only *mostly* optional, but the SQL language standard is decided by a standards committee made up of industry competitors).

**Comment by paul1956** [comment]
With VB 16 you can already do
```VB
dim result = SyntaxFactory.MultiLineIfBlock(ifStatement)
                .WithStatements(GenerateNewTrueStatement(ifBlock.ElseBlock.Statements)) _ ' This _ might be optional
_ '                .WithLeadingTrivia(ifBlock.GetLeadingTrivia())
_ '                .WithTrailingTrivia(ifBlock.GetTrailingTrivia())
                .WithAdditionalAnnotations(Formatter.Annotation)
```

**Comment by paul1956** [comment]
There is currently a design review about how the _ on a new line should be handled. I think it should be aligned with or 2 spaces before the . on the line above vs how it is shown above. There is a WIP very ugly PR against Roslyn showing options and current behavior.

**Comment by pricerc** [comment]
@paul1956 . 

I understand it's a tricky proposition. But with few specific exceptions, T-SQL's parser manages to combine "rest-of-line" comments with statement continuation on the next line, without the need for statement delimiters or line-continuation characters.

So we know it can be done. 

And I think it would be cool if VB could lose (especially) line continuation characters.

Unfortunately, I don't have enough clock cycles at the moment to look at it myself, because I'm really curious about the inner workings of the compiler. Sadly running a new startup means that I don't have time for such intense non-billable work at the moment.

**Comment by paul1956** [comment]
I don't think SQL Parser has the VB "With" Statement and object initializes, how would you know if the "." was part of the With/Initializer or continued from the last line.  I added _ ' to allow comments almost anywhere with is very useful for language translation without losing comments and formatting. It allows comments on blank lines without terminating the previous statement. Changing the parser is relatively easy writing all the tests to prove you didn't break anything is a lot of work and doesn't interest me.

**Comment by pricerc** [comment]
> I don't think SQL Parser has the VB "With" Statement and object initializes, how would you know if the "." was part of the With/Initializer or continued from the last line. 

Curiously. A SQL query starting with `WITH` requires the preceding statement to be terminated with a semi-colon...

disclaimer: late-night possibly confused and contradictory observations ahead.  Just spit-balling.

To answer your question. It's complicated. But you already knew that.

In my ideal world `With` wouldn't exist.  Even for initializers. Problem solved.

Firstly, many years ago, I discovered that I couldn't edit-and-continue when using `With`. And so I stopped. And since none of the other languages I use has the construct, I just use the same technique I use everywhere else - use a short acronym and just repeat it over several lines. For all I know, it works now, but I'm not about to relearn something I have hardly any use for, and a workaround that I've grown to prefer. But I digress.

Regarding initializers. Why do we need an explicit initializer? 

So, if I'm instantiating an object, and then setting a bunch of properties, FxCop and/or CodeRush want me to convert that to an initializer. Why not just do that with the compiler? 

why should
```VB
dim x = new SomeObject() 
x.Foo = Bar
```
compile differently than
```VB
dim x = new SomeObject() with { .Foo = Bar }
```
?

Then dealing with `With ... End With`.

All the cool kids are doing Fluent stuff, right?

so (trivial example):
```VB
With x = new StringBuilder("foo").Append("bar")
  .Append("baz")
  .Replace("baz", "bar")
End With
```
could just as easily be:
```VB
dim x = new StringBuilder("foo").Append("bar")
  .Append("baz")
  .Replace("baz", "bar")
```

As I'm typing this, I'm recognising the irony: VB's `With` has been providing a Fluent-ish paradigm for years. It was just spoilt by requiring the words `With` and `End With`.

I need to think about this some more when my brain is working better.

Peace out.

**Comment by paul1956** [comment]
@pricerc love your realization. I use With everywhere it make code very readable and keeps lines short.
I would never write because lines become way to long
```
Dim x = new SomeObject() with { .Foo = Bar }
```
It would write below with appropriate indentation because it is very clear what is happening and lines stay short
```
Dim x = new SomeObject() with {
                             .Foo = Bar
                             }
```
The IDE has a  CodeFix that will convert If I miss one. It is possible the compiler could do it since the CodeFix does but the issue for me is readability and short lines and the compiler doing it doesn't solve either of those. I assume would hope both forms generate the same code.

**Comment by pricerc** [comment]
> @pricerc love your realization. 

Because of the nature of my work, I don't get as much time as I'd like to philosophise on the tools I'm using. So it sometimes takes me a while to figure out something that's bothering me about the latest cool trend in software.

So it is with '_fluent_' interfaces. 

When I first started seeing them, I found myself wondering what problem the solution was for. And I've never been fully comfortable with it.  I found a (php-focused) article that I think articulates some of my concerns well : [fluent-interfaces-are-evil](https://ocramius.github.io/blog/fluent-interfaces-are-evil/]) (despite the name, the author includes positives as well).

As I was typing last night, I realized just how much a `with` statement looks like a chain of _fluent_ methods, and found myself wondering if _fluent_ and `with` originated with similar goals.

Then having ignored a few conversations today while considering this, I find myself wondering if there's a case for combining the existing capabilities of `with` and _fluent_ concepts to produce something more powerful.

I'm going to have to come back to this later; it'll take me hours to think it through and articulate it properly, with a fair bit of digging through this forum at past conversations.

**Comment by jrmoreno1** [comment]
@paul1956 : are you saying this feature has been implemented, or is your example badly formatted (has . starting line).

**Comment by paul1956** [comment]
@jrmoreno1 which example? Below work in VB 16 which is included in VS 16.X (not sure when language default was changed to 16 from 15)
```VB
    Function Test() As Integer
        Dim i As Integer _ ' this is a
 _ ' Another comment
 _ '
            = 1
        Return i
    End Function
```

**Comment by paul1956** [comment]
If SomeObject exists then the formatting is reasonable
```
    Public Shared Function Test() As Integer
        Dim x As New IO.FileInfo("C:\temp") With
            {
            .IsReadOnly = True
        } ' This does not intent as I would do it
        Return 1
    End Function
```
